### ISSUE TITLE
Is there a way to convert request_animation_frame() to Promise?

### ISSUE THREAD

**Issue by ababo** [rustwasm/wasm-bindgen#2574]
I need to call `request_animation_frame()` as async function. Can you share a code to wrap `request_animation_frame()` into `Promise`/`Future`?

**Comment by ababo** [comment]
I think I figured that out, closing.

**Comment by nouzu** [comment]
> I think I figured that out, closing.

Please post your solution.

**Comment by daxpedda** [comment]
You can just use [`wasm_bindgen_futures::spawn_local()`](https://docs.rs/wasm-bindgen-futures/latest/wasm_bindgen_futures/fn.spawn_local.html) inside the passed function.

Let us know if you need any further help.

**Comment by nouzu** [comment]
> You can just use [`wasm_bindgen_futures::spawn_local()`](https://docs.rs/wasm-bindgen-futures/latest/wasm_bindgen_futures/fn.spawn_local.html) inside the passed function.
> 
> Let us know if you need any further help.

I don't see how `wasm_bindgen_futures::spawn_local()` is helpful. I need to `await` on `request_animation_frame` instead of doing [what I'm currently doing](https://github.com/nouzu/visual_sort/blob/791f47e80e7d88ef29c7d40c6ea8de25fd79840e/modules/renderer/src/lib.rs#L290).

**Comment by daxpedda** [comment]
> I need to `await` on `request_animation_frame` instead of doing [what I'm currently doing](https://github.com/nouzu/visual_sort/blob/791f47e80e7d88ef29c7d40c6ea8de25fd79840e/modules/renderer/src/lib.rs#L290).

You can just `await` inside `spawn_local()`, I'm not sure what the difference is. `await`ing inside `requestAnimationFrame` won't do anything special. For example it won't prevent a new `requestAnimationFrame` to trigger, because you have to queue that anyway.

Is there some side-effect of `await`ing in `requestAnimationFrame` I'm missing?

**Comment by daxpedda** [comment]
To answer your question though, you can return a `Promise` in the function you are passing to `requestAnimationFrame`, that is basically the same thing as `requestAnimationFrame(async () => console.log("hello"))`.

**Comment by nouzu** [comment]
> You can just `await` inside `spawn_local()`, I'm not sure what the difference is. `await`ing inside `requestAnimationFrame` won't do anything special. For example it won't prevent a new `requestAnimationFrame` to trigger, because you have to queue that anyway.
> 
> Is there some side-effect of `await`ing in `requestAnimationFrame` I'm missing?

I don't need to `await` in `spawn_local` nor in `request_animation_frame`, I need to `await` the next animation frame (`window.request_animation_frame().await)`, so that I can write my logic sequentially, without callbacks.

I've tried the following (not sure if that's what you're suggesting)
```rust
#[wasm_bindgen]
pub async fn shuffle() {
    let state = unsafe { STATE.as_mut().unwrap() };

    loop {
        state.array.shuffle(&mut thread_rng());

        state.render_array();

        // wait for next frame?
        {
            let (s, r) = oneshot::channel();
            wasm_bindgen_futures::spawn_local(async { s.send(()).unwrap(); });
            r.await.unwrap();
        }
    }
}
```
but it just freezes the page, probably because the scheduler behind `spawn_local` isn't using `request_animation_frame`,

**Comment by daxpedda** [comment]
Ah! I misunderstood the original problem.
Unfortunately this isn't possible, because `wasm-bindgen` is just wrapping the JS Web API and `requestAnimationFrame` just doesn't return a `Promise` or anything you can `await`: https://developer.mozilla.org/en-US/docs/Web/API/window/requestAnimationFrame.

The code you posted above probably freezes the pages because it will just loop endlessly, nothing is actually stopping it or waiting on anything, `spawn_local` happens immediately after the context switch.

What you want to do, is just something like this (simplifying some stuff here):
```rust
async fn wait_for_frame() {
	let flag = Flag::new();
	{
		let flag = flag.clone();
		window.request_animation_frame(Closure::once(move || flag.raise()));
	}

	flag.await;
} 
```

Hope this helps!

**Comment by nouzu** [comment]
```rust
async fn request_animation_frame() {
    let (s, r) = oneshot::channel();

    let closure = Closure::once(move || s.send(()).unwrap());

    web_sys::window().unwrap()
        .request_animation_frame(closure.as_ref().unchecked_ref()).unwrap();

    r.await.unwrap();
}
```

An obvious solution and yet the monkey brain wasn't able to generate it until this instant. Thanks for the help @daxpedda.
